### PR TITLE
Improve docstring of base tuner and assessor

### PR DESCRIPTION
**Pull request by liuzhe-lz**
Rewrite tuner and assessor docstring.

Currently `load_checkpoint`, `save_checkpoint`, and `import_data` are marked as internal APIs. Because as an internal developer I can't figure out how to use them myself.

There are some links to rst/md doc in docstring. I don't know how to add external links. So I simply wrote the file path.

And seems there is a less verbose way to add cross-reference, tell me if you know.

This PR also hides `Tuner.accept_customized_trials` from public interface. This API is likely to change if we refactor advisor interface in next release. (We can safely assume nobody is using it because customized trial is not supported yet.)

### CODE REVIEW

File: src/sdk/pynni/nni/assessor.py [hunk 1]
```diff
@@ -18,44 +18,118 @@
...
+    """The trial works poorly and should be early stopped."""
+
 
 class Assessor(Recoverable):
+    """
+    Assessor analyzes trial's intermediate results (e.g., periodically evaluated accuracy on test dataset)
+    to tell whether this trial can be early stopped or not.
+
+    This is the abstract base class for all assessors.
+    Early stopping algorithms should derive this class and override :meth:`assess_trial` method,
```
Comment (ultmaster) [inline src/sdk/pynni/nni/assessor.py]: Same here.

File: src/sdk/pynni/nni/assessor.py [hunk 2]
```diff
@@ -18,44 +18,118 @@
...
+    which receives intermediate results from trials and give an assessing result.
+
+    If :meth:`assess_trial` returns :obj:`AssessResult.Bad` for a trial,
+    it hints NNI framework that the trial is likely to result in a poor final accuracy,
+    and therefore should be killed to save resource.
+
+    If an accessor want's to get notified when a trial ends, it can also override :meth:`trial_end`.
```
Comment (ultmaster) [inline src/sdk/pynni/nni/assessor.py]: wants to be notified

File: src/sdk/pynni/nni/assessor.py [hunk 3]
```diff
@@ -18,44 +18,118 @@
...
-        """Determines whether a trial should be killed. Must override.
-        trial_job_id: identifier of the trial (str).
-        trial_history: a list of intermediate result objects.
-        Returns AssessResult.Good or AssessResult.Bad.
+        """
+        Abstract method for determining whether a trial should be killed. Must override.
+
+        The NNI framework has little guarantee on ``trial_history``.
+        This method is not guaranteed to be invoked for each time ``trial_history`` get updated.
+        It is also possible that a trial's history keeps updateing after receiving a bad result.
```
Comment (ultmaster) [inline src/sdk/pynni/nni/assessor.py]: updating

File: src/sdk/pynni/nni/assessor.py [hunk 4]
```diff
@@ -18,44 +18,118 @@
...
-        success: True if the trial successfully completed; False if failed or terminated.
+        """
+        Abstract method invoked when a trial is completed or terminated. Do nothing by default.
+
+        Parameters
+        ----------
+        trial_job_id: str
+            Unique identifier of the trial.
+        success: bool
+            True if the trial successfully completed; False if failed or terminated.
```
Comment (ultmaster) [inline src/sdk/pynni/nni/assessor.py]: Maybe I don't know about Python. Shouldn't there be a `pass` here?

File: src/sdk/pynni/nni/tuner.py [hunk 5]
```diff
@@ -17,31 +17,128 @@
...
 from .recoverable import Recoverable
 
+__all__ = ['Tuner']
+
 _logger = logging.getLogger(__name__)
 
 
 class Tuner(Recoverable):
+    """
+    Tuner is an AutoML algorithm, which generates a new configuration for the next try.
```
Comment (ultmaster) [inline src/sdk/pynni/nni/tuner.py]: generates configurations to run with.

File: src/sdk/pynni/nni/tuner.py [hunk 6]
```diff
@@ -17,31 +17,128 @@
...
+    A new trial will run with this configuration.
+
+    This is the abstract base class for all tuners.
+    Tuning algorithms should derive this class and override :meth:`update_search_space`, :meth:`receive_trial_result`,
+    as well as :meth:`generate_parameters` or :meth:`generate_multiple_parameters`.
+
+    After initializing, NNI will first call :meth:`update_search_space` to tell tuner the feasible region,
+    and then call :meth:`generate_parameters` one or more times to request for hyper-parameter configurations.
```
Comment (ultmaster) [inline src/sdk/pynni/nni/tuner.py]: Better mention `generate_parameter` here.

**Comment by ultmaster** [comment]
Oops. Seems PR has been merged. Please ignore my comments. (No big problem here)

File: src/sdk/pynni/nni/tuner.py [hunk 7]
```diff
@@ -17,31 +17,119 @@
...
 _logger = logging.getLogger(__name__)
 
 
 class Tuner(Recoverable):
+    """
+    Tuner is an AutoML algorithm, which generates a new configuration for the next try.
+    A new trial will run with this configuration.
+
+    This is the abstract base class for all tuners.
+    Tuning algorithms should derive this class and override :meth:`update_search_space`, :meth:`receive_trial_result`,
```
[the remark this answers is not in the document]
Comment (QuanluZhang) [reply src/sdk/pynni/nni/tuner.py]: @ultmaster  merged here, please create another pr for your comments

File: src/sdk/pynni/nni/assessor.py (hunk 4 above)
Comment (liuzhe-lz) [reply src/sdk/pynni/nni/assessor.py]: Pylint with defaut configuration will complain if there is a `pass`.
I tried to find some PEP or other offical doc to clarify whether there shoud be a `pass` or not, but failed.

File: src/sdk/pynni/nni/tuner.py (hunk 5 above)
Comment (liuzhe-lz) [reply src/sdk/pynni/nni/tuner.py]: This sentence is copied from [overview](https://github.com/microsoft/nni/blob/master/docs/en_US/Overview.md).
Do you think it should be updated?

File: src/sdk/pynni/nni/tuner.py (hunk 6 above)
Comment (liuzhe-lz) [reply src/sdk/pynni/nni/tuner.py]: Do you mean `generate_parameters`? It's already there.

File: src/sdk/pynni/nni/assessor.py (hunk 4 above)
Comment (ultmaster) [reply src/sdk/pynni/nni/assessor.py]: If Pylint didn't complain, I'm fine with it.

File: src/sdk/pynni/nni/tuner.py (hunk 5 above)
Comment (ultmaster) [reply src/sdk/pynni/nni/tuner.py]: Not sure. "For the next try" looks really Chi-english to me.

File: src/sdk/pynni/nni/tuner.py (hunk 6 above)
Comment (ultmaster) [reply src/sdk/pynni/nni/tuner.py]: I mean `generate_multiple_parameters`. I think users should have a general idea about which method (multiple or not multiple) to override at the very beginning. You could move the explanation in `gen_mul_param` to here. Putting it in tutorials is better, but... never mind.

**Pull request merged**